Mathlib/Algebra/Ring/BooleanRing.lean

```diff
@@ -0,0 +1,624 @@
+/-
+Copyright (c) 2021 Bryan Gin-ge Chen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bryan Gin-ge Chen, Yaël Dillies
+
+! This file was ported from Lean 3 source module algebra.ring.boolean_ring
+! leanprover-community/mathlib commit e8638a0fcaf73e4500469f368ef9494e495099b3
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.PUnitInstances
+import Mathlib.Tactic.Abel
+import Mathlib.Tactic.Ring
+import Mathlib.Order.Hom.Lattice
+import Mathlib.Tactic.ScopedNS
+
+/-!
+# Boolean rings
+
+A Boolean ring is a ring where multiplication is idempotent. They are equivalent to Boolean
+algebras.
+
+## Main declarations
+
+* `BooleanRing`: a typeclass for rings where multiplication is idempotent.
+* `BooleanRing.toBooleanAlgebra`: Turn a Boolean ring into a Boolean algebra.
+* `BooleanAlgebra.ToBooleanRing`: Turn a Boolean algebra into a Boolean ring.
+* `AsBoolAlg`: Type-synonym for the Boolean algebra associated to a Boolean ring.
+* `AsBoolRing`: Type-synonym for the Boolean ring associated to a Boolean algebra.
+
+## Implementation notes
+
+We provide two ways of turning a Boolean algebra/ring into a Boolean ring/algebra:
+* Instances on the same type accessible in locales `BooleanAlgebraOfBooleanRing` and
+  `BooleanRingOfBooleanAlgebra`.
+* Type-synonyms `AsBoolAlg` and `AsBoolRing`.
+
+At this point in time, it is not clear the first way is useful, but we keep it for educational
+purposes and because it is easier than dealing with
+`ofBoolAlg`/`toBoolAlg`/`ofBoolRing`/`toBoolRing` explicitly.
+
+## Tags
+
+boolean ring, boolean algebra
+-/
+
+
+variable {α β γ : Type _}
+
+/-- A Boolean ring is a ring where multiplication is idempotent. -/
+class BooleanRing (α) extends Ring α where
+  /-- Multiplication in a boolean ring is idempotent. -/
+  mul_self : ∀ a : α, a * a = a
+#align boolean_ring BooleanRing
+
+section BooleanRing
+
+variable [BooleanRing α] (a b : α)
+
+instance : IsIdempotent α (· * ·) :=
+  ⟨BooleanRing.mul_self⟩
+
+-- Porting note: simpNF linter complains because this lemmas in other files not to be in simpNF
```

Comment:
```suggestion
-- Porting note: simpNF linter complains because this causes other lemmas in other files not to be in simpNF
```
and below.

Reply:
Also note that it does *not*. The lemmas in other files aren't about `boolean_ring`, so in full generality they can't be rewritten using it. Understand, the linter is wrong here.

Reply:
The porting notes still wrongly claims that those lemmas make other simp lemmas not simp normal form!